### PR TITLE
Configure queue for async requests

### PR DESCRIPTION
**Pull request by alekseysemenov**
@ctapobep

### CODE REVIEW

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java [hunk 1]
```diff
@@ -34,16 +34,15 @@
...
-import java.util.concurrent.Future;
+import java.util.concurrent.*;
 
 @Singleton
 @Named(ArtifactUpdateApiClientImpl.ID)
 public class ArtifactUpdateApiClientImpl extends ComponentSupport implements ArtifactUpdateApiClient {
 
     public static final String ID = "artifactUpdateApiClient";
+    public static final Integer REQUESTS_SENDING_THREADS_COUNT = 1;
+    public static final Integer REQUESTS_QUEUE_SIZE = 10;
```
Comment (kevstigneev) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java]: 10 -> 100

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java [hunk 2]
```diff
@@ -51,9 +50,17 @@
     private ConfigurationsManager configurationsManager;
 
     /**
+     * Queue for async requests
+     */
+    private BlockingQueue<Runnable> queue = new LinkedBlockingQueue<>(REQUESTS_QUEUE_SIZE);
```
Comment (kevstigneev) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java]: _Blocking_ queue?
> A Queue that ..., and **wait for space to become available** in the queue when storing an element. 

It blocks event processor, doesn't it?

File: replication-plugin.xml [hunk 3]
```diff
@@ -13,8 +13,11 @@ User and password are optional. If there is neither login nor password (or login
 You could also define myUrl attribute for <configurations> tag that defines URL of the Nexus instance.
 Remote URL of the proxy repository (at the peer instance) should start with myUrl (configured for master instance).
 This configuration used to match what proxy repository have to poll artifact.
+
+requestsQueueSize and requestsSendingThreadsCount attributes defines thread pool configurations. This pool is used to execute
```
Comment (ctapobep) [inline replication-plugin.xml]: defines -> define
configurations -> configuration
execute -> send

Comment (ctapobep) [inline replication-plugin.xml]: Also, I'd put defaults here in the docs.

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java [hunk 4]
```diff
@@ -51,13 +48,23 @@
     private ConfigurationsManager configurationsManager;
```
Comment (ctapobep) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java]: `final`?

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java [hunk 5]
```diff
@@ -51,13 +48,23 @@
...
 
     /**
+     * Queue for async requests
+     */
+    private BlockingQueue<Runnable> queue;
```
Comment (ctapobep) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java]: Field can be converted into a local var.

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java [hunk 6]
```diff
@@ -51,13 +48,23 @@
...
+
+    /**
      * ExecutorService shares between clients. All treads are created in the same executor
      */
-    private ExecutorService asyncRequestsExecutorService = Executors.newFixedThreadPool(10);
+    private ExecutorService asyncRequestsExecutorService;
```
Comment (ctapobep) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java]: `final`?

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java [hunk 7]
```diff
@@ -51,13 +48,23 @@
...
 
     @Inject
     public ArtifactUpdateApiClientImpl(ConfigurationsManager configurationsManager) {
         this.configurationsManager = configurationsManager;
+        this.queue = new LinkedBlockingQueue<>(configurationsManager.getConfiguration().getRequestsQueueSize());
+        this.asyncRequestsExecutorService = new ThreadPoolExecutor(
+                configurationsManager.getConfiguration().getRequestsSendingThreadsCount(),
+                configurationsManager.getConfiguration().getRequestsSendingThreadsCount(),
+                30, TimeUnit.SECONDS, queue);
```
Comment (ctapobep) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java]: - 30 secs seems to be too low, WDYT?
- The default Rejection Execution Handler is the one that simply throws exception. Is that what we _actually_ want when there is no place for a new request? If so, I'd document this explicitly - the decision itself and the reason why we chose it.

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/UploadEventListenerImpl.java [hunk 8]
```diff
@@ -73,23 +73,23 @@ public void onArtifactUploading(RepositoryItemEventStore event) {
                 ArtifactMetaInfo metaInfo = new ArtifactMetaInfo(configurationsManager.getConfiguration().getMyUrl(), gav.getGroupId(), gav.getArtifactId(), gav.getVersion(), repo.getId());
                 metaInfo.setClassifier(gav.getClassifier());
                 metaInfo.setExtension(gav.getExtension());
-                ArtifactStatus artifactStatus = getArtifactStatus(metaInfo);
+                ArtifactStatus artifactStatus = null;
                 if (!gav.isSignature() && !gav.isHash()) {
```
Comment (ctapobep) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/UploadEventListenerImpl.java]: This method definitely requires tests. The logic can be factored out into a sep method and a unit test can be written for it.

**Comment by gd-tmagrys** [comment]
I agree with all remarks from @ctapobep. To see if I have any additional remarks, first we need to fix those. @kevstigneev you'd like me to fix this remarks ?

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/UploadEventListenerImpl.java (hunk 8 above)
Comment (gd-tmagrys) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/UploadEventListenerImpl.java]: I agree. This method definitely needs to be tested by unit tests.

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java (hunk 7 above)
Comment (gd-tmagrys) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java]: What do you think about using unbounded blocking queue ?

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/UploadEventListenerImpl.java (hunk 8 above)
Comment (gd-tmagrys) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/UploadEventListenerImpl.java]: We agree with @kevstigneev to write tests for this method while implementing issue "Automated tests" after release 1.0

**Comment by kevstigneev** [comment]
This PR is overriden by #27

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java (hunk 7 above)
Comment (ctapobep) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java]: Well, you know the problem with the unbounded queues - they are unbounded :) Stability of Nexus should be considered as a priority while unbounded queues is a risk to get OOM.

**Pull request closed without merging**